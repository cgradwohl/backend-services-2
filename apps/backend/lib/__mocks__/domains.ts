export const addTenantToDomain = jest.fn();
export const getDomain = jest.fn();
