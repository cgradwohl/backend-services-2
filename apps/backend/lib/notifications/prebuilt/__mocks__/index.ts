export const getWelcomePrebuiltTemplate = jest.fn();
