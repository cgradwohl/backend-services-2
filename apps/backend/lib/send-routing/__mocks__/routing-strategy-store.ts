export const getSendRoutingStrategy = jest.fn();
