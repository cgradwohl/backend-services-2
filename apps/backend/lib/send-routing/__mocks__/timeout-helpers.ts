export const isRouteLeafTimedOut = jest.fn();
export const setSendTimesForLeafs = jest.fn(() => ({
  message: "2022-07-14T21:53:16.495Z",
  channels: {
    sms: "2022-07-14T21:53:16.495Z",
  },
}));
