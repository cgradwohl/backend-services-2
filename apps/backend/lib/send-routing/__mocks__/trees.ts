import { RouteNode } from "../types";

// Generated from (with some manual tweaks to rep edge case)
// routing: {
//   method: "single",
//   channels: ["sms", "email"],
// },
export const complexSmsAndEmailTree: RouteNode = {
  address: [],
  nodes: [
    {
      address: [0],
      nodes: [
        {
          channel: "sms",
          provider: "twilio",
          providerConfigurationId: "7593c6b6-f241-47f0-85b4-b5386cd60086",
          taxonomy: "direct_message:sms:twilio",
          address: [0, 0],
          providerFailoverIndex: 1,
          type: "leaf",
        },
        {
          failureReason: "unknown",
          failureType: "FILTERED",
          channel: "sms",
          provider: "shrug",
          address: [0, 1],
          type: "dead-leaf",
        },
      ],
      type: "branch",
      failover: {
        address: [0, "failover"],
        nodes: [
          {
            channel: "sms",
            provider: "vonage",
            providerConfigurationId: "7593c6b6-f241-47f0-85b4-vonage",
            taxonomy: "direct_message:sms:vonage",
            address: [0, "failover", 0],
            providerFailoverIndex: 2,
            type: "leaf",
          },
        ],
        type: "branch",
        failover: {
          address: [0, "failover", "failover"],
          // // To make this extra tricky we make this a send to "all" strategy
          nodes: [
            {
              channel: "sms",
              provider: "sinch",
              providerConfigurationId: "7593c6b6-f241-47f0-85b4-sinch",
              taxonomy: "direct_message:sms:sinch",
              address: [0, "failover", "failover", 0],
              providerFailoverIndex: 3,
              type: "leaf",
            },
            {
              address: [0, "failover", "failover", 1],
              nodes: [
                {
                  channel: "sms",
                  provider: "abc",
                  providerConfigurationId: "7593c6b6-f241-47f0-85b4-abc",
                  taxonomy: "direct_message:sms:abc",
                  address: [0, "failover", "failover", 1, 0],
                  providerFailoverIndex: 3,
                  type: "leaf",
                },
              ],
              type: "branch",
              failover: {
                address: [0, "failover", "failover", 1, "failover"],
                nodes: [
                  {
                    channel: "sms",
                    provider: "123",
                    providerConfigurationId: "7593c6b6-f241-47f0-85b4-123",
                    taxonomy: "direct_message:sms:123",
                    address: [0, "failover", "failover", 1, "failover", 0],
                    providerFailoverIndex: 4,
                    type: "leaf",
                  },
                ],
                type: "branch",
              },
            },
          ],
          type: "branch",
        },
      },
    },
  ],
  type: "branch",
  failover: {
    address: ["failover"],
    nodes: [
      {
        address: ["failover", 0],
        nodes: [
          {
            channel: "email",
            provider: "mailjet",
            providerConfigurationId: "7593c6b6-f241-47f0-85b4-b5386cd600er6",
            taxonomy: "email:mailjet",
            address: ["failover", 0, 0],
            providerFailoverIndex: 1,
            type: "leaf",
          },
        ],
        type: "branch",
        failover: {
          address: ["failover", 0, "failover"],
          nodes: [
            {
              channel: "email",
              provider: "sendgrid",
              providerConfigurationId: "2321",
              taxonomy: "email:sendgrid",
              address: ["failover", 0, "failover", 0],
              providerFailoverIndex: 2,
              type: "leaf",
            },
          ],
          type: "branch",
        },
      },
    ],
    type: "branch",
  },
};
