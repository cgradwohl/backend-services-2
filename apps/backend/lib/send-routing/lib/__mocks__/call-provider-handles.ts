const { callProviderHandles: actualCallProviderHandles } = jest.requireActual(
  "~/lib/send-routing/lib/call-provider-handles"
);
export const callProviderHandles = jest.fn(actualCallProviderHandles);
