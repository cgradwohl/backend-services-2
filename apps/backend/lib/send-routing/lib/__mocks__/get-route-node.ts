export const getRouteNode = jest.fn(() => ({ type: "leaf" }));
