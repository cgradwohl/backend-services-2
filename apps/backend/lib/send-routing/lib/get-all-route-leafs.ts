import { DeadRouteBranch, DeadRouteLeaf, RouteLeaf, RouteNode } from "../types";

export interface GetAllRouteLeafsOpts {
  includeFailoverBranches?: boolean;
  includeDeadLeafs?: boolean;
}

export interface GetAllRouteLeafsWithDeadLeafsOpts
  extends GetAllRouteLeafsOpts {
  includeDeadLeafs: true;
}

export function getAllRouteLeafs(
  tree: RouteNode,
  opts: GetAllRouteLeafsWithDeadLeafsOpts
): (RouteLeaf | DeadRouteLeaf | DeadRouteBranch)[];
export function getAllRouteLeafs(
  tree: RouteNode,
  opts?: GetAllRouteLeafsOpts
): RouteLeaf[];
export function getAllRouteLeafs(
  tree: RouteNode,
  opts?: GetAllRouteLeafsOpts
): (RouteLeaf | DeadRouteLeaf | DeadRouteBranch)[] {
  if (tree.type === "branch") {
    return [
      ...tree.nodes.flatMap((n) => getAllRouteLeafs(n, opts)),
      ...(opts?.includeFailoverBranches && tree.failover
        ? getAllRouteLeafs(tree.failover, opts)
        : []),
    ];
  }

  if (tree.type === "leaf") {
    return [tree];
  }

  if (
    opts?.includeDeadLeafs &&
    (tree.type === "dead-leaf" || tree.type === "dead-branch")
  ) {
    return [tree];
  }

  return [];
}
