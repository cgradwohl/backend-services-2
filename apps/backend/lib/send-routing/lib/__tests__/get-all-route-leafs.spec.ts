import { complexSmsAndEmailTree } from "../../__mocks__/trees";
import { getAllRouteLeafs } from "../get-all-route-leafs";

describe("get all route leafs", () => {
  it("should return only live non-failover leafs", () => {
    expect(getAllRouteLeafs(complexSmsAndEmailTree)).toMatchSnapshot();
  });

  it("should return all live leafs including from failover branches", () => {
    expect(
      getAllRouteLeafs(complexSmsAndEmailTree, {
        includeFailoverBranches: true,
      })
    ).toMatchSnapshot();
  });

  it("should return all live and dead leafs", () => {
    expect(
      getAllRouteLeafs(complexSmsAndEmailTree, {
        includeDeadLeafs: true,
      })
    ).toMatchSnapshot();
  });
});
