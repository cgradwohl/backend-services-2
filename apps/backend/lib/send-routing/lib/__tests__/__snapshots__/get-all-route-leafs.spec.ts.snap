// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get all route leafs should return all live and dead leafs 1`] = `
Array [
  Object {
    "address": Array [
      0,
      0,
    ],
    "channel": "sms",
    "provider": "twilio",
    "providerConfigurationId": "7593c6b6-f241-47f0-85b4-b5386cd60086",
    "providerFailoverIndex": 1,
    "taxonomy": "direct_message:sms:twilio",
    "type": "leaf",
  },
  Object {
    "address": Array [
      0,
      1,
    ],
    "channel": "sms",
    "failureReason": "unknown",
    "failureType": "FILTERED",
    "provider": "shrug",
    "type": "dead-leaf",
  },
]
`;

exports[`get all route leafs should return all live leafs including from failover branches 1`] = `
Array [
  Object {
    "address": Array [
      0,
      0,
    ],
    "channel": "sms",
    "provider": "twilio",
    "providerConfigurationId": "7593c6b6-f241-47f0-85b4-b5386cd60086",
    "providerFailoverIndex": 1,
    "taxonomy": "direct_message:sms:twilio",
    "type": "leaf",
  },
  Object {
    "address": Array [
      0,
      "failover",
      0,
    ],
    "channel": "sms",
    "provider": "vonage",
    "providerConfigurationId": "7593c6b6-f241-47f0-85b4-vonage",
    "providerFailoverIndex": 2,
    "taxonomy": "direct_message:sms:vonage",
    "type": "leaf",
  },
  Object {
    "address": Array [
      0,
      "failover",
      "failover",
      0,
    ],
    "channel": "sms",
    "provider": "sinch",
    "providerConfigurationId": "7593c6b6-f241-47f0-85b4-sinch",
    "providerFailoverIndex": 3,
    "taxonomy": "direct_message:sms:sinch",
    "type": "leaf",
  },
  Object {
    "address": Array [
      0,
      "failover",
      "failover",
      1,
      0,
    ],
    "channel": "sms",
    "provider": "abc",
    "providerConfigurationId": "7593c6b6-f241-47f0-85b4-abc",
    "providerFailoverIndex": 3,
    "taxonomy": "direct_message:sms:abc",
    "type": "leaf",
  },
  Object {
    "address": Array [
      0,
      "failover",
      "failover",
      1,
      "failover",
      0,
    ],
    "channel": "sms",
    "provider": "123",
    "providerConfigurationId": "7593c6b6-f241-47f0-85b4-123",
    "providerFailoverIndex": 4,
    "taxonomy": "direct_message:sms:123",
    "type": "leaf",
  },
  Object {
    "address": Array [
      "failover",
      0,
      0,
    ],
    "channel": "email",
    "provider": "mailjet",
    "providerConfigurationId": "7593c6b6-f241-47f0-85b4-b5386cd600er6",
    "providerFailoverIndex": 1,
    "taxonomy": "email:mailjet",
    "type": "leaf",
  },
  Object {
    "address": Array [
      "failover",
      0,
      "failover",
      0,
    ],
    "channel": "email",
    "provider": "sendgrid",
    "providerConfigurationId": "2321",
    "providerFailoverIndex": 2,
    "taxonomy": "email:sendgrid",
    "type": "leaf",
  },
]
`;

exports[`get all route leafs should return only live non-failover leafs 1`] = `
Array [
  Object {
    "address": Array [
      0,
      0,
    ],
    "channel": "sms",
    "provider": "twilio",
    "providerConfigurationId": "7593c6b6-f241-47f0-85b4-b5386cd60086",
    "providerFailoverIndex": 1,
    "taxonomy": "direct_message:sms:twilio",
    "type": "leaf",
  },
]
`;
